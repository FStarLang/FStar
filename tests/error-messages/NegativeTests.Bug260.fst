(*
   Copyright 2008-2018 Microsoft Research

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)
module NegativeTests.Bug260
type pnat =
  | O : pnat
  | S : pnat -> pnat

type validity : n:pnat -> Type =
  | VSucc : n:pnat -> Tot (validity (S n))
val bad : t:pnat -> Tot (validity (S (S t)))

[@@(expect_failure [19])]
let bad t = VSucc t


(* Hard to keep this one in the suite since the program fails to check even with --admit_smt_queries true *)
(* Update 2024/06/29: This now works and seems totally fine. *)

assume val good : int -> Type
assume val enc: plain:int -> c:unit{good plain}
assume val repr : int -> int

let f (text:int) = enc (repr text)