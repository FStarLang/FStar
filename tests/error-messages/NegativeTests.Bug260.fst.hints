[
  "’Ë×¡0³\u0012vº«ŒïuS!M",
  [
    [
      "NegativeTests.Bug260.pnat",
      1,
      2,
      1,
      [ "@query" ],
      0,
      "f8df6099ce967b050496809e0338d9b4"
    ],
    [
      "NegativeTests.Bug260.__proj__S__item___0",
      1,
      2,
      1,
      [
        "@MaxIFuel_assumption", "@query",
        "refinement_interpretation_Tm_refine_c7798980fdaa69b62544817f36c3402f"
      ],
      0,
      "095109921a3791cb8fda6fb7f81fbcf3"
    ],
    [
      "NegativeTests.Bug260.validity",
      1,
      2,
      1,
      [ "@query", "assumption_NegativeTests.Bug260.pnat__uu___haseq" ],
      0,
      "7419834845299b08bc26415b13533789"
    ]
  ]
]