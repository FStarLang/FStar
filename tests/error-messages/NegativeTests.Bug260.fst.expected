>> Got issues: [
NegativeTests.Bug260.fst(26,12-26,19): (Error 19) Subtyping check failed; expected type validity (S (S t)); got type validity (S t); The SMT solver could not prove the query. Use --query_stats for more details. (see also NegativeTests.Bug260.fst(23,37-26,9))
>>]
NegativeTests.Bug260.fst(23,4-23,7): (Warning 240) Admitting NegativeTests.Bug260.bad without a definition
Verified module: NegativeTests.Bug260
All verification conditions discharged successfully
